ons-spark/pyspark-intro/reading-data-pyspark.ipynb

```diff
@@ -0,0 +1,358 @@
+      " |-- AnimalGroupParent: string (nullable = true)\n",
+      " |-- OriginofCall: string (nullable = true)\n",
+      " |-- PropertyType: string (nullable = true)\n",
+      " |-- PropertyCategory: string (nullable = true)\n",
+      " |-- SpecialServiceTypeCategory: string (nullable = true)\n",
+      " |-- SpecialServiceType: string (nullable = true)\n",
+      " |-- WardCode: string (nullable = true)\n",
+      " |-- Ward: string (nullable = true)\n",
+      " |-- BoroughCode: string (nullable = true)\n",
+      " |-- Borough: string (nullable = true)\n",
+      " |-- StnGroundName: string (nullable = true)\n",
+      " |-- PostcodeDistrict: string (nullable = true)\n",
+      " |-- Easting_m: double (nullable = true)\n",
+      " |-- Northing_m: double (nullable = true)\n",
+      " |-- Easting_rounded: integer (nullable = true)\n",
+      " |-- Northing_rounded: integer (nullable = true)\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "animal_rescue.printSchema()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To correct this, we can either cast the column as a date type or we can provide a schema when reading it in. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "root\n",
+      " |-- IncidentNumber: string (nullable = true)\n",
+      " |-- DateTimeOfCall: timestamp (nullable = true)\n",
+      " |-- CalYear: integer (nullable = true)\n",
+      " |-- TypeOfIncident: string (nullable = true)\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "from pyspark.sql.types import StructType, StructField, StringType, IntegerType, DoubleType, TimestampType\n",
+    "\n",
+    "custom_schema = StructType([\n",
+    "    StructField(\"IncidentNumber\", StringType(), True),\n",
+    "    StructField(\"DateTimeOfCall\", TimestampType(), True),\n",
+    "    StructField(\"CalYear\", IntegerType(), True),\n",
+    "    StructField(\"TypeOfIncident\", StringType(), True),])\n",
+    "\n",
+    "animal_rescue = spark.read.csv(path = config[\"rescue_path_csv\"], header = True, schema = custom_schema)\n",
+    "animal_rescue.printSchema()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can see that DateTimeOfCall has now been read in correctly as a timestamp type. Also, providing a schema will improve the efficiency of the operation. This is because, in order to infer a schema, Spark needs to scan the dataset. A column could contain, for example, an integer for the first 1000 rows and a string for the 1001th row - in which case it would be inferred as a string, but this wouldn't be obvious from the first 1000 rows. Needing to sample the data in each column can be quite memory intensive. Providing a schema means that Spark no longer has to sample the data before reading it in. "
```

Comment:
Here it is! I hadn't thought about the point that scanning is memory intensive, I suppose that is true. What I remember from my training is that it takes time and will be repeated with repeated actions, but fair point about resource. It's just unnecessary resource being used isn't it.

Reply:
Added in a section about it taking longer and being repeated every action.